poetapp/node
# Warning on Blockchain Fork

[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

Part of https://github.com/poetapp/node/issues/46
Fixes https://github.com/poetapp/node/issues/647

## How to Test

Add the following to `docker-compose.override.yml`:

```yml
  bitcoind-2:
    image: ruimarinho/bitcoin-core:0.17.0
    ports:
      - "28443:18443"
    command:
      -printtoconsole
      -server
      -regtest=1
      -rpcuser=bitcoinrpcuser
      -rpcpassword=bitcoinrpcpassword
      -rpcallowip=::/0
      -deprecatedrpc=signrawtransaction
      -connect=bitcoind-1
    links:
    - bitcoind-1
    depends_on:
    - bitcoind-1
```

We're using port `28443` for the RPC so it doesn't conflict with the port `18443` that was already taken by `bitcoind-1`.

- Generate some blocks in `bitcoind-1`: `bitcoin-cli -rpcport=18443 generate 3`.
- Start `bitcoind-2`: `docker-compose up -d bitcoind-2`.
- Now run `bitcoin-cli -rpcport=28443 getblockchaininfo` and `bitcoin-cli -rpcport=18443 getblockchaininfo` — both commands should output similar results after a few seconds. This means `bitcoind-1` sent all the blocks it had in it to `bitcoind-2` and both nodes see the same blockchain.
- `npm start`. The Po.et node is reading from `bitcoind-1`.
- `bitcoin-cli -rpcport=18443 setnetworkactive false`
- At this stage, `bitcoin-cli getchaintips` should output the same for both bitcoin nodes.
- `bitcoin-cli -rpcport=18443 generate 1`
- `bitcoin-cli -rpcport=28443 generate 1`
- Now, `bitcoin-cli getchaintips` will output different data for either bitcoin node. Both should have a single `active` branch of the same `height` but with different `hash`.
- `bitcoin-cli -rpcport=28443 generate 1`
- Now, `bitcoin-cli getchaintips` will show a longer chain for `bitcoind-2` than `bitcoind-2`.
- `bitcoin-cli -rpcport=18443 setnetworkactive true`
- At this moment the two bitcoin nodes will reconnect and share the blocks they have. Since both chains are valid, but the chain from `bitcoind-2` is longer, `bitcoind-1` will discard its last block and add the two last blocks from `bitcoind-2`.
- Running `bitcoin-cli getchaintips` will now output two elements: one `active` and one `valid-fork`. Both bitcoin nodes should have the same `active` chain, and only `bitcoind-1` should have the `valid-fork`.
- A `Fork detected` warning should have been logged by the node.
